## Machine Learning - Practical/Workshop 1 - Data Manipulation and Subset Selection

### Preliminaries

**Aim**: The aim of this workshop is to familiarize yourself with some basic concepts and data science tools in R, before exploring methods for model selection. You have already used R in Intro to Stats. Here we will introduce you to other tools in R that are frequently used in data science.

**R software**: Here you will work with R via the Durham University cluster in Computer Science. If you want to use R on your own machine, we recommend Rstudio (you probably already installed it during Intro to Stats) or a Jupyter Notebook with an R kernel via Anaconda. You can also use other services such as Google Colab (https://colab.research.google.com/), GitHub Codespaces (https://github.com/features/codespaces).

### Part I - Working with Data


Datasets in R are often stored in special types of lists called dataframes (similar to tables in Excel and Pandas in Python). Dataframes are matrix-like blocks of values where each column represents a variable, and each row represents a case. Unlike generic lists, each sublist must contain the same number of elements.

Say we want to create a table with the top 5 grossing films in the UK since 1989 (not corrected for inflation) based on the data in [Wikipedia](https://en.wikipedia.org/wiki/List_of_highest-grossing_films_in_the_United_Kingdom).

1. The column headers for our initial table will be **Rank** and **Title**.
2. The values in **Rank** are the numbers 1 to 5.
3. The values in **Title** are Star Wars:The Force Awakens, Skyfall, No Time to Die, Spider-Man: No Way Home, and Avatar.

In [5]:
movies = data.frame(
  Rank = 1:5,
  Title = c("Star Wars: The Force Awakens",
            "Skyfall",
            "No Time to Die",
            "Spiderman: No Way Home",
            "Avatar")
)
View(movies)

Rank Title                       
1 1    Star Wars: The Force Awakens
2 2    Skyfall                     
3 3    No Time to Die              
4 4    Spiderman: No Way Home      
5 5    Avatar

Take a look at the [R Data Frames page](https://www.w3schools.com/r/r_data_frames.asp) in W3 Schools for more examples.

In the snippet above, we use the View function instead of print. If using RStudio, View opens the dataset in a new tab that let's you preview your dataset, and filter and sort your data more easily. See [View function in R](https://www.statology.org/view-function-in-r/).

Let's add two more columns to the dataframe to include the **Gross** in millions of pounds to the column after **Title**, and then the **Year** of release between the Title and Gross.

In [6]:
movies['Gross']=c(123.3, 102.8, 98.0, 97.2, 96.7)
View(movies)

Rank Title                        Gross
1 1    Star Wars: The Force Awakens 123.3
2 2    Skyfall                      102.8
3 3    No Time to Die                98.0
4 4    Spiderman: No Way Home        97.2
5 5    Avatar                        96.7

In [7]:
movies['Year']=c(2015,2012,2021,2021,2009)
movies=movies[,c('Rank','Title','Year','Gross')]
View(movies)

Rank Title                        Year Gross
1 1    Star Wars: The Force Awakens 2015 123.3
2 2    Skyfall                      2012 102.8
3 3    No Time to Die               2021  98.0
4 4    Spiderman: No Way Home       2021  97.2
5 5    Avatar                       2009  96.7

There are many ways of adding new rows to a dataframe. If you only need to add one row to the end, you can do as below:

In [8]:
movies[6,]=c(6, "Barbie", 2023, 95.6)
View(movies)

Rank Title                        Year Gross
1 1    Star Wars: The Force Awakens 2015 123.3
2 2    Skyfall                      2012 102.8
3 3    No Time to Die               2021 98   
4 4    Spiderman: No Way Home       2021 97.2 
5 5    Avatar                       2009 96.7 
6 6    Barbie                       2023 95.6

If you need to add multiple rows, you should create another dataset with the same columns as the original and **rbind** it to the previous one. Later on we will introduce you to the **tidyverse** which will give you more possibilities to execute tasks like this.

In [9]:
movies_newrows = data.frame(
  Rank = 7:8,
  Title = c("Spectre","Avengers:Endgame"),
  Year = c(2015, 2019),
  Gross = c(95.2, 88.7)
)
movies = rbind(movies, movies_newrows)
View(movies)

Rank Title                        Year Gross
1 1    Star Wars: The Force Awakens 2015 123.3
2 2    Skyfall                      2012 102.8
3 3    No Time to Die               2021 98   
4 4    Spiderman: No Way Home       2021 97.2 
5 5    Avatar                       2009 96.7 
6 6    Barbie                       2023 95.6 
7 7    Spectre                      2015 95.2 
8 8    Avengers:Endgame             2019 88.7

Now let's sort this dataset by **Year** in descending order. To do so without introducing new packages, we use the **order** function, not the sort function! The sort function returns the entries in sorted order but it doesn't work well with dataframes; the order function returns the indices of the sorted entries.

By default sorting is ascending using the **order** function so you need to set the *decreasing* parameter to TRUE.

In [10]:
View(
  movies[order(movies$Year,decreasing=TRUE),]
)

Rank Title                        Year Gross
6 6    Barbie                       2023 95.6 
3 3    No Time to Die               2021 98   
4 4    Spiderman: No Way Home       2021 97.2 
8 8    Avengers:Endgame             2019 88.7 
1 1    Star Wars: The Force Awakens 2015 123.3
7 7    Spectre                      2015 95.2 
2 2    Skyfall                      2012 102.8
5 5    Avatar                       2009 96.7

#### **Exercise 1**

Add the next 4 movies to the dataset and sort it by year in **ascending** order:

- Top Gun: Maverick - 2022 - 83.7
- Star Wars: The Last Jedi - 2017 - 82.7
- Titanic - 1998 - 82.7
- Avatar: The Way of Water - 2022 - 76.9

### Dataframes and Matrices

Let's load the **Boston** dataset from the **MASS** library. Check your Intro to Stats workshops as a refresher on how to load libraries and datasets!

The Boston dataset contains information on multiple attributes for suburbs in Boston, Massachusetts [Boston in R Documentation](https://stat.ethz.ch/R-manual/R-devel/library/MASS/html/Boston.html).



In [38]:
library(MASS)
data(Boston)
head(Boston, n=5) #The head function shows the first n rows in a dataframe

crim    zn indus chas nox   rm    age  dis    rad tax ptratio black  lstat
1 0.00632 18 2.31  0    0.538 6.575 65.2 4.0900 1   296 15.3    396.90 4.98 
2 0.02731  0 7.07  0    0.469 6.421 78.9 4.9671 2   242 17.8    396.90 9.14 
3 0.02729  0 7.07  0    0.469 7.185 61.1 4.9671 2   242 17.8    392.83 4.03 
4 0.03237  0 2.18  0    0.458 6.998 45.8 6.0622 3   222 18.7    394.63 2.94 
5 0.06905  0 2.18  0    0.458 7.147 54.2 6.0622 3   222 18.7    396.90 5.33 
  medv
1 24.0
2 21.6
3 34.7
4 33.4
5 36.2

Most of the well-known machine learning models are implemented in R (and Python!), and are relatively easy to use. However, there will be cases where you'll need to prepare and transform your data using techniques seen elsewhere on the program (in Intro to Maths, for example, where you learned how to manipulate matrices and vectors, invert matrices, and solve systems of linear equations). There might also be cases where you want to implement your own version of a machine learning model and modify it; therefore, it is useful to know how to manipulate datasets and matrices in R (and understand how to move between those two data types!).

In Intro to Stats Week 6, you studied linear regression models (check your notes!), and you were trying to fit the model:

$Y=X\beta+\epsilon$

and estimate the vector $\beta$.  
To estimate $\beta$ using least squares, we want to compute:

$\hat{\beta}=(X^TX)^{-1}X^TY$

where $X^T$ is the transpose of $X$ and $A^{-1}$ represents the inverse of a matrix $A$.

#### **Briefly review Intro to Stats workshops 6 and 7 before continuing**

You can also try your solutions here in this notebook to get familiar with the environment.

Now, let's start with a simple example using linear regression and the Boston dataset.



#### **Exercise 2** - Fitting a linear regression model with **lm**

Recall that you used the **lm** function to fit linear models in Intro to Stats.  Use the **lm** function here to fit a linear regression model with *medv* as the response variable and the following variables as the predictors: *rm*, *lstat*, *indus*, and *ptratio*.

Save the estimated values of the regression coefficients to the vector **b_lm**.

Now let's try to calculate $\hat\beta$ using matrix operations via
$\hat{\beta}=(X^TX)^{-1}X^TY$.

To do that, we have to define $X$ and $Y$. $Y$ is the vector containing the values of *medv*. The $X$ matrix is the *design matrix* composed by a unit vector linked to the model intercept ($\beta_0$), and the vectors containing the values of the variables *rm*, *lstat*, *indus*, and *ptratio*.

Let's create Y and X:

In [3]:
Y = Boston$medv
Xpart = Boston[,c('rm','lstat','indus','ptratio')]
X = cbind(1,Xpart) #this looks like cheating but it works for dataframes,
#you can also use rep to first create a unit vector!
head(X)

1 rm    lstat indus ptratio
1 1 6.575 4.98  2.31  15.3   
2 1 6.421 9.14  7.07  17.8   
3 1 7.185 4.03  7.07  17.8   
4 1 6.998 2.94  2.18  18.7   
5 1 7.147 5.33  2.18  18.7   
6 1 6.430 5.21  2.18  18.7

Now let's convert $Y$ and $X$ to matrices using the *as.matrix* function.

In [4]:
Y = as.matrix(Y)
X = as.matrix(X)

To transpose the matrix $X$, we use the **t** function. To invert it, we use the **solve** function. Or if you really want to practice your coding skills, you can implement your own inversion function later on!

To multiply matrices and vectors in R, we use %*%.

In [5]:
trX = t(X)
trXX = trX%*%X
trXX

1        rm        lstat     indus     ptratio  
1        506.000  3180.025   6402.45   5635.21   9338.50
rm      3180.025 20234.598  38681.79  34461.82  58415.97
lstat   6402.450 38681.788 106762.96  86240.71 121080.91
indus   5635.210 34461.816  86240.71  86525.63 106875.32
ptratio 9338.500 58415.973 121080.91 106875.32 174713.93

In [6]:
inv_trXX = solve(trXX)
inv_trXX

1             rm            lstat         indus         ptratio      
1        0.5627402385 -5.212583e-02 -2.636070e-03  4.386852e-04 -1.109171e-02
rm      -0.0521258284  6.632218e-03  3.659520e-04 -2.220860e-06  3.163865e-04
lstat   -0.0026360697  3.659520e-04  8.368636e-05 -3.579378e-05 -1.755929e-05
indus    0.0004386852 -2.220860e-06 -3.579378e-05  6.935899e-05 -4.032735e-05
ptratio -0.0110917066  3.163865e-04 -1.755929e-05 -4.032735e-05  5.296313e-04

#### **Exercise 3**

Continue the calculations above and return a vector of length 5, assigning it the name **b_calc**. Compare your vector to **b_lm**.

#### **Exercise 4**

Write down and execute the steps you would take to calculate the coefficients for

*lm(medv~poly(lstat,2, raw=TRUE), data=Boston)*

using matrices as above.

What does the *raw=TRUE* option do in the *poly* function?
Increase the power of the polynomial and check what happens to your matrices.

Note: Remember, you can treat powers of *lstat* as separate feature variables

### PART II - Best Subset Selection

**Preparation**: Take a look at the contents for weeks 7 and 8 in Introduction to Statistics. We will build on some of the concepts you have already seen such as variable selection and model validation.

#### **Introduction**: 

Above, we used the *Boston* dataset to fit a linear model using **lm**. We had *medv* as our response variable, and the following variables as our predictors: *rm*, *lstat*, *indus*, and *ptratio*.

In [9]:
library(MASS)
data(Boston)
summary(lm(medv~rm+lstat+indus+ptratio, data=Boston))


Call:
lm(formula = medv ~ rm + lstat + indus + ptratio, data = Boston)

Residuals:
     Min       1Q   Median       3Q      Max 
-14.5602  -3.1379  -0.7984   1.7783  29.5739 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 18.614970   3.926680   4.741 2.78e-06 ***
rm           4.515179   0.426286  10.592  < 2e-16 ***
lstat       -0.575711   0.047885 -12.023  < 2e-16 ***
indus        0.007567   0.043594   0.174    0.862    
ptratio     -0.935122   0.120464  -7.763 4.71e-14 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 5.234 on 501 degrees of freedom
Multiple R-squared:  0.6786,	Adjusted R-squared:  0.6761 
F-statistic: 264.5 on 4 and 501 DF,  p-value: < 2.2e-16


We can see that the $R^2$ for this model is 0.6786. This means that 67.86% of the variation in the quantitative measure of property median value (medv) can be explained by its linear regression on the 4 chosen predictor variables. We can also see that the adjusted $R^2$ is 0.6761.

Say we now include the *nox* predictor in our model:

In [10]:
summary(lm(medv~rm+lstat+indus+ptratio+nox, data=Boston))


Call:
lm(formula = medv ~ rm + lstat + indus + ptratio + nox, data = Boston)

Residuals:
    Min      1Q  Median      3Q     Max 
-14.111  -3.078  -0.831   1.748  29.924 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 20.60281    4.16490   4.947 1.03e-06 ***
rm           4.58538    0.42871  10.696  < 2e-16 ***
lstat       -0.55338    0.05035 -10.990  < 2e-16 ***
indus        0.06188    0.05796   1.068    0.286    
ptratio     -0.97272    0.12322  -7.894 1.86e-14 ***
nox         -4.72784    3.32874  -1.420    0.156    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 5.229 on 500 degrees of freedom
Multiple R-squared:  0.6799,	Adjusted R-squared:  0.6767 
F-statistic: 212.4 on 5 and 500 DF,  p-value: < 2.2e-16


The $R^2$ has moved from 67.61% to 67.99%. So, if we assume that a model with a higher $R^2$ is better, we would choose this model over the previous one. In this case, the adjusted $R^2$ is also higher at 0.6767.

**Question:** Which measure should we use to compare the first and the second model? $R^2$ or the adjusted $R^2$? Discuss this with a colleague or a tutor before proceeding.


SImilarly, if we add the variable *zn* to the first model, we will see a decrease in the adjusted $R^2$ in comparison to the first model. We can also see a decrease in the $R^2$ in comparison to the second model. Therefore, we would (possibly) conclude that we are better off with the second model.

In [11]:
summary(lm(medv~rm+lstat+indus+ptratio+zn, data=Boston))


Call:
lm(formula = medv ~ rm + lstat + indus + ptratio + zn, data = Boston)

Residuals:
     Min       1Q   Median       3Q      Max 
-14.5903  -3.1327  -0.8342   1.8571  29.5849 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 19.148274   3.996972   4.791 2.19e-06 ***
rm           4.527117   0.426807  10.607  < 2e-16 ***
lstat       -0.577861   0.048000 -12.039  < 2e-16 ***
indus       -0.004404   0.046642  -0.094    0.925    
ptratio     -0.953931   0.123288  -7.737 5.65e-14 ***
zn          -0.008859   0.012233  -0.724    0.469    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 5.237 on 500 degrees of freedom
Multiple R-squared:  0.679,	Adjusted R-squared:  0.6758 
F-statistic: 211.5 on 5 and 500 DF,  p-value: < 2.2e-16


If we are to consider all possible linear combinations of variables, we would need to consider the model with no predictors (only an intercept) and the model with all predictors available. Let's see what the calculated $R^2$ and adjusted $R^2$ is for these models:

In [12]:
m0=lm(medv~1, data=Boston)
summary(m0)


Call:
lm(formula = medv ~ 1, data = Boston)

Residuals:
    Min      1Q  Median      3Q     Max 
-17.533  -5.508  -1.333   2.467  27.467 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  22.5328     0.4089   55.11   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 9.197 on 505 degrees of freedom


In [13]:
mfull=lm(medv~ . , data=Boston)
summary(mfull)


Call:
lm(formula = medv ~ ., data = Boston)

Residuals:
    Min      1Q  Median      3Q     Max 
-15.595  -2.730  -0.518   1.777  26.199 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  3.646e+01  5.103e+00   7.144 3.28e-12 ***
crim        -1.080e-01  3.286e-02  -3.287 0.001087 ** 
zn           4.642e-02  1.373e-02   3.382 0.000778 ***
indus        2.056e-02  6.150e-02   0.334 0.738288    
chas         2.687e+00  8.616e-01   3.118 0.001925 ** 
nox         -1.777e+01  3.820e+00  -4.651 4.25e-06 ***
rm           3.810e+00  4.179e-01   9.116  < 2e-16 ***
age          6.922e-04  1.321e-02   0.052 0.958229    
dis         -1.476e+00  1.995e-01  -7.398 6.01e-13 ***
rad          3.060e-01  6.635e-02   4.613 5.07e-06 ***
tax         -1.233e-02  3.760e-03  -3.280 0.001112 ** 
ptratio     -9.527e-01  1.308e-01  -7.283 1.31e-12 ***
black        9.312e-03  2.686e-03   3.467 0.000573 ***
lstat       -5.248e-01  5.072e-02 -10.347  < 2e-16 ***
---
Signif. codes:  0

**Question**: What happened to the $R^2$ and adjusted $R^2$ for the model $M_0$ above? Again discuss this with your colleagues or a tutor.

#### **Best Subset Selection procedure**

To perform best subset selection, we fit a separate regression model for each possible combination of the $p$ predictors. This is often broken up into stages, as follows:

1. Let   $M_0$ denote the model which contains no predictors.

2. For $k=1,2,\ldots,p$:

- Fit all $p \choose k$  models that contain exactly $k$ predictors.
- Pick the best among these
  models and call it $M_k$. Here, best is defined as having the smallest RSS or largest $R^2$.

3. Select a single best model from $M_0,M_1,\ldots,M_p$ using a suitable measure such as the cross-validated prediction error $C_p$, BIC, adjusted $R^2$, etc.

It is important to note that use of RSS or $R^2$ in step 2 of the above algorithm is acceptable because the models all have an equal number of predictors. We can't use RSS in step 3 because RSS decreases monotonically as the number of predictors included in the model increases.

**Exercise**: Let's write a loop to try to find the best subset model for *medv*.

*Note*: There are many different ways to complete this exercise using different libraries, data subsetting/manipulation, etc. The way we are going to attempt this now is not optimal but it is simple and easy to implement.

**Step 1:** Say we want to fit the model *medv ~ nox*, we write:

In [15]:
m_example=lm(medv ~ nox, data=Boston)
summary(m_example)


Call:
lm(formula = medv ~ nox, data = Boston)

Residuals:
    Min      1Q  Median      3Q     Max 
-13.691  -5.121  -2.161   2.959  31.310 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   41.346      1.811   22.83   <2e-16 ***
nox          -33.916      3.196  -10.61   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 8.323 on 504 degrees of freedom
Multiple R-squared:  0.1826,	Adjusted R-squared:  0.181 
F-statistic: 112.6 on 1 and 504 DF,  p-value: < 2.2e-16


**Step 2:** And we can get the $R^2$ and adjusted $R^2$ for this model by calling:

In [16]:
r2_example = summary(m_example)$r.squared
adjr2_example = summary(m_example)$adj.r.squared
print(paste0("R2 = ",r2_example))
print(paste0("Adj R2 = ",adjr2_example))

[1] "R2 = 0.182603042501699"
[1] "Adj R2 = 0.180981223141583"


**Step 3:** When we called the model with all variables in the previous section, we used *lm(medv ~ ., data=Boston)*. The dot indicates we want to use all variables in the dataset Boston.

Instead of writing *lm(medv ~ nox, data=Boston)* in the previous model, we could write *lm(medv ~ ., data=Boston[,c(5,14)])*.

The 5th column in Boston corresponds to the variable *nox* and the last column (14th) corresponds to *medv*. So we have subsetted the dataset keeping only the columns of interest for this model.

In [8]:
summary(lm(medv ~ ., data=Boston[,c(5,14)]))


Call:
lm(formula = medv ~ ., data = Boston[, c(5, 14)])

Residuals:
    Min      1Q  Median      3Q     Max 
-13.691  -5.121  -2.161   2.959  31.310 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   41.346      1.811   22.83   <2e-16 ***
nox          -33.916      3.196  -10.61   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 8.323 on 504 degrees of freedom
Multiple R-squared:  0.1826,	Adjusted R-squared:  0.181 
F-statistic: 112.6 on 1 and 504 DF,  p-value: < 2.2e-16


**Step 4:** Putting everything together, we can write a simple loop to compare all models containing one variable, and saving their $R^2$ to a vector.  We also print some statements that inform us of the most relevant information from our calculations in an easy-to-read way.

In [17]:
r2_m1 = rep(NA,13)
for (var in 1:13){
  m_temp = lm(medv~ . , data = Boston[,c(var,14)])
  r2_m1[var]=summary(m_temp)$r.squared
}
print(r2_m1)
print(paste0("The maximum calcuated R2 is: ", round( max(r2_m1), 6 )))  # look at the help file for round.
print(paste0("The index of the corresponding model is: ",which.max(r2_m1)))
print(paste0("The relevant predictor is: ", names(Boston)[which.max(r2_m1)]))

 [1] 0.15078047 0.12992084 0.23399003 0.03071613 0.18260304 0.48352546
 [7] 0.14209474 0.06246437 0.14563858 0.21952592 0.25784732 0.11119612
[13] 0.54414630
[1] "The maximum calcuated R2 is: 0.544146"
[1] "The index of the corresponding model is: 13"
[1] "The relevant predictor is: lstat"


**Step 5:** Say we now want to look at all models with 2 variables, we would have to find all pairwise combinations of predictors. To do this, we can use the function *combn*:

In [18]:
all_pairs = combn(13,2)
all_pairs
dim(all_pairs)[2]

[,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]
[1,] 1    1    1    1    1    1    1    1     1    1    ⋯      9     9     9   
[2,] 2    3    4    5    6    7    8    9    10   11    ⋯     10    11    12   
     [,15] [,16] [,17] [,18] [,19] [,20] [,21]
[1,]  9    10    10    10    11    11    12   
[2,] 13    11    12    13    12    13    13

[1] 78

**Step 6:** And we can use the same principle to subset the Boston dataset as before to generate the model using the variables in column 1 of *all_pairs*:

In [19]:
summary(lm(medv ~ ., data = Boston[,c(all_pairs[,1],14)]))


Call:
lm(formula = medv ~ ., data = Boston[, c(all_pairs[, 1], 14)])

Residuals:
    Min      1Q  Median      3Q     Max 
-15.421  -5.060  -1.558   2.121  30.765 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 22.48563    0.44173  50.904  < 2e-16 ***
crim        -0.35208    0.04259  -8.267 1.24e-15 ***
zn           0.11611    0.01571   7.392 6.09e-13 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 8.065 on 503 degrees of freedom
Multiple R-squared:  0.234,	Adjusted R-squared:  0.2309 
F-statistic: 76.82 on 2 and 503 DF,  p-value: < 2.2e-16


**Step 7:** Now write a loop similar to the one in **Step 4** and find the model with the highest $R^2$ with two predictors.

**Step 8:** Repeat steps 5 to 7 and create a loop that returns the model with the highest value of $R^2$ for $1, 2, \ldots, 13$ variables. Find a strategy to save the index for the relevant variables in each model, the corresponding $R^2$, and the adjusted $R^2$ for each.

**Step 9:** Compare the adjusted $R^2$ for the models in **Step 8** and identify the best subset model.

### **Painless best subset selection**

Now that you have seen one way to implement subset selection, you should be ready to use functions that do this job for you without the need to implement your own loops.

We use the library **leaps** to return the best subset model.

In [22]:
library(leaps)
best_models = regsubsets(medv ~ ., data=Boston)
summary(best_models)

Subset selection object
Call: regsubsets.formula(medv ~ ., data = Boston, nvmax = 13)
13 Variables  (and intercept)
        Forced in Forced out
crim        FALSE      FALSE
zn          FALSE      FALSE
indus       FALSE      FALSE
chas        FALSE      FALSE
nox         FALSE      FALSE
rm          FALSE      FALSE
age         FALSE      FALSE
dis         FALSE      FALSE
rad         FALSE      FALSE
tax         FALSE      FALSE
ptratio     FALSE      FALSE
black       FALSE      FALSE
lstat       FALSE      FALSE
1 subsets of each size up to 13
Selection Algorithm: exhaustive
          crim zn  indus chas nox rm  age dis rad tax ptratio black lstat
1  ( 1 )  " "  " " " "   " "  " " " " " " " " " " " " " "     " "   "*"  
2  ( 1 )  " "  " " " "   " "  " " "*" " " " " " " " " " "     " "   "*"  
3  ( 1 )  " "  " " " "   " "  " " "*" " " " " " " " " "*"     " "   "*"  
4  ( 1 )  " "  " " " "   " "  " " "*" " " "*" " " " " "*"     " "   "*"  
5  ( 1 )  " "  " " " "   " "  "*" "*" " " "*

The *regsubsets* function returns 8 possible "best subset" models.  To return more (or less) models, you can change the argument *nvmax* (by default set to 8) within the summary function.  Therefore, to check the best model with 1 up to 13 variables, we run: 

In [26]:
best_models = regsubsets(medv ~ ., data=Boston,  nvmax = 13)
summary(best_models)

Subset selection object
Call: regsubsets.formula(medv ~ ., data = Boston, nvmax = 13)
13 Variables  (and intercept)
        Forced in Forced out
crim        FALSE      FALSE
zn          FALSE      FALSE
indus       FALSE      FALSE
chas        FALSE      FALSE
nox         FALSE      FALSE
rm          FALSE      FALSE
age         FALSE      FALSE
dis         FALSE      FALSE
rad         FALSE      FALSE
tax         FALSE      FALSE
ptratio     FALSE      FALSE
black       FALSE      FALSE
lstat       FALSE      FALSE
1 subsets of each size up to 13
Selection Algorithm: exhaustive
          crim zn  indus chas nox rm  age dis rad tax ptratio black lstat
1  ( 1 )  " "  " " " "   " "  " " " " " " " " " " " " " "     " "   "*"  
2  ( 1 )  " "  " " " "   " "  " " "*" " " " " " " " " " "     " "   "*"  
3  ( 1 )  " "  " " " "   " "  " " "*" " " " " " " " " "*"     " "   "*"  
4  ( 1 )  " "  " " " "   " "  " " "*" " " "*" " " " " "*"     " "   "*"  
5  ( 1 )  " "  " " " "   " "  "*" "*" " " "*

To check the values that we can obtain from this summary object, you can use the *names* function:

In [23]:
res_summary = summary(best_models)
names(res_summary)

[1] "which"  "rsq"    "rss"    "adjr2"  "cp"     "bic"    "outmat" "obj"

From which we can see that we can obtain the adjusted $R^2$ values for the models using the *adjr2* component of our *summary* object: 

In [24]:
res_summary$adjr2

[1] 0.5432418 0.6371245 0.6767036 0.6878351 0.7051702 0.7123567 0.7182560
 [8] 0.7222072 0.7252743 0.7299149 0.7348058 0.7343282 0.7337897

and to identify the model with the highest adjusted $R^2$, we use the function **which.max**:

In [25]:
which.max(res_summary$adjr2)

[1] 11

The coefficients for model 11 are then given by:

In [27]:
print(coef(best_models,11))

  (Intercept)          crim            zn          chas           nox 
 36.341145004  -0.108413345   0.045844929   2.718716303 -17.376023429 
           rm           dis           rad           tax       ptratio 
  3.801578840  -1.492711460   0.299608454  -0.011777973  -0.946524570 
        black         lstat 
  0.009290845  -0.522553457 


We can construct the linear model corresponding to these variable as follows using:

In [28]:
lm(medv~crim+zn+chas+nox+rm+dis+rad+tax+ptratio+black+lstat, data=Boston)


Call:
lm(formula = medv ~ crim + zn + chas + nox + rm + dis + rad + 
    tax + ptratio + black + lstat, data = Boston)

Coefficients:
(Intercept)         crim           zn         chas          nox           rm  
  36.341145    -0.108413     0.045845     2.718716   -17.376023     3.801579  
        dis          rad          tax      ptratio        black        lstat  
  -1.492711     0.299608    -0.011778    -0.946525     0.009291    -0.522553  


Alternatively, we can generate this linear model by reformulating the coefficient names in the best model into an expression that can be read by *lm*.  To do this, notice that we can call the names of the coefficients in the best model using *names*:

In [29]:
names(coef(best_models,11))

[1] "(Intercept)" "crim"        "zn"          "chas"        "nox"        
 [6] "rm"          "dis"         "rad"         "tax"         "ptratio"    
[11] "black"       "lstat"

We can then create a string expression of the model formula we require using *paste*, where we remove the first element as we do not wish to include the word _intercept_ within our model expression formula, and then link the words using _+_:

In [30]:
paste(names(coef(best_models,11))[-1],collapse="+")

[1] "crim+zn+chas+nox+rm+dis+rad+tax+ptratio+black+lstat"

We can then use this expression to *reformulate* the required linear model expression with response *medv*, with the argument *data* being used within *lm* as usual:

In [31]:
best_ss <- lm(reformulate(paste(names(coef(best_models,11))[-1],collapse="+"),'medv'),data=Boston)
summary( best_ss )


Call:
lm(formula = reformulate(paste(names(coef(best_models, 11))[-1], 
    collapse = "+"), "medv"), data = Boston)

Residuals:
     Min       1Q   Median       3Q      Max 
-15.5984  -2.7386  -0.5046   1.7273  26.2373 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  36.341145   5.067492   7.171 2.73e-12 ***
crim         -0.108413   0.032779  -3.307 0.001010 ** 
zn            0.045845   0.013523   3.390 0.000754 ***
chas          2.718716   0.854240   3.183 0.001551 ** 
nox         -17.376023   3.535243  -4.915 1.21e-06 ***
rm            3.801579   0.406316   9.356  < 2e-16 ***
dis          -1.492711   0.185731  -8.037 6.84e-15 ***
rad           0.299608   0.063402   4.726 3.00e-06 ***
tax          -0.011778   0.003372  -3.493 0.000521 ***
ptratio      -0.946525   0.129066  -7.334 9.24e-13 ***
black         0.009291   0.002674   3.475 0.000557 ***
lstat        -0.522553   0.047424 -11.019  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 

### Forwards and Backwards Subset Selection

Forwards and backwards subset selection are similar techniques which do not make use of every possible combination of variables.

**Forward selection**

In the case of forwards subset selection, we start with the null model, and determine which of the $p$ single-predictor models has the highest $R^2$ value. We designate this model $M_1$.

Next, rather than fitting each of the $p \choose 2$ two-predictor models, we add each of the remaining $p-1$ predictor variables to $M_1$ and determine which of these models has the highest $R^2$ value. We then designate this model $M_2$.  We then generate $p-2$ models with three variables by adding each of the remaining $p-2$ variables to the variables included in $M_2$.  The model with highest $R^2$ value is denoted $M_3$.

We continue in this fashion until we have our set of $p$ models $M_0,M_1,\ldots,M_p$, from which we choose the "best" model using $C_p$, BIC, adjusted $R^2$, etc.

**Backward selection**

In the case of backwards subset selection, we start from the *full* model $M_p$ and *remove* a single variable at a time.

The model from this set of $p$ models which has the highest $R^2$ value is designated $M_{p-1}$. We continue in this fashion until we reach the null model, $M_0$. We again choose the best model from $M_0,M_1,\ldots,M_p$ based on $C_p$, BIC, adjusted $R^2$, etc.  


#### Implementation using Leaps

We can use the **leaps** package and *regsubsets* to conduct forward and backward selection, we just need to specify the method. To perform forward selection and select the best model using the adjust $R^2$, we include our method choice in the function call as follows.

In [34]:
library(leaps)
best_forward = regsubsets(medv ~ ., data=Boston, method="forward", nvmax = 13)
summary(best_forward)

Subset selection object
Call: regsubsets.formula(medv ~ ., data = Boston, method = "forward", 
    nvmax = 13)
13 Variables  (and intercept)
        Forced in Forced out
crim        FALSE      FALSE
zn          FALSE      FALSE
indus       FALSE      FALSE
chas        FALSE      FALSE
nox         FALSE      FALSE
rm          FALSE      FALSE
age         FALSE      FALSE
dis         FALSE      FALSE
rad         FALSE      FALSE
tax         FALSE      FALSE
ptratio     FALSE      FALSE
black       FALSE      FALSE
lstat       FALSE      FALSE
1 subsets of each size up to 13
Selection Algorithm: forward
          crim zn  indus chas nox rm  age dis rad tax ptratio black lstat
1  ( 1 )  " "  " " " "   " "  " " " " " " " " " " " " " "     " "   "*"  
2  ( 1 )  " "  " " " "   " "  " " "*" " " " " " " " " " "     " "   "*"  
3  ( 1 )  " "  " " " "   " "  " " "*" " " " " " " " " "*"     " "   "*"  
4  ( 1 )  " "  " " " "   " "  " " "*" " " "*" " " " " "*"     " "   "*"  
5  ( 1 )  " "  " " " "

In [35]:
forward_summary = summary(best_forward)
forward_adjr2=which.max(forward_summary$adjr2)
print(coef(best_forward,forward_adjr2))

  (Intercept)          crim            zn          chas           nox 
 36.341145004  -0.108413345   0.045844929   2.718716303 -17.376023429 
           rm           dis           rad           tax       ptratio 
  3.801578840  -1.492711460   0.299608454  -0.011777973  -0.946524570 
        black         lstat 
  0.009290845  -0.522553457 


In [36]:
summary(lm(reformulate(paste(names(coef(best_forward,forward_adjr2))[-1],collapse="+"),'medv'),data=Boston))


Call:
lm(formula = reformulate(paste(names(coef(best_forward, forward_adjr2))[-1], 
    collapse = "+"), "medv"), data = Boston)

Residuals:
     Min       1Q   Median       3Q      Max 
-15.5984  -2.7386  -0.5046   1.7273  26.2373 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  36.341145   5.067492   7.171 2.73e-12 ***
crim         -0.108413   0.032779  -3.307 0.001010 ** 
zn            0.045845   0.013523   3.390 0.000754 ***
chas          2.718716   0.854240   3.183 0.001551 ** 
nox         -17.376023   3.535243  -4.915 1.21e-06 ***
rm            3.801579   0.406316   9.356  < 2e-16 ***
dis          -1.492711   0.185731  -8.037 6.84e-15 ***
rad           0.299608   0.063402   4.726 3.00e-06 ***
tax          -0.011778   0.003372  -3.493 0.000521 ***
ptratio      -0.946525   0.129066  -7.334 9.24e-13 ***
black         0.009291   0.002674   3.475 0.000557 ***
lstat        -0.522553   0.047424 -11.019  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘*

In this case, the forward selection process leads to the same model as the one obtained using best subset selection.

**Exercise:** Repeat the process above using *backward* as the method option. How different is this model from the best subset and forward models?